pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,907 @@
+	}
+
+	// provisioning is not complete, return Pending
+	if !isAnnotationMatch(pvProvisioningRequired, pvProvisioningCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] - provisioning in progress", pv.Name)
+		// one lock per volume to allow parallel processing but limit activity per volume
+		if _, exists := controller.locks[pv.Name]; !exists {
+			controller.locks[pv.Name] = &sync.RWMutex{}
+			go provisionVolume(pv, controller)
+		}
+
+		return pv, api.PersistentVolumeStatus{Phase: api.VolumePending, Message: "Awaiting provisioning"}, nil
+	}
+
+	return pv, api.PersistentVolumeStatus{Phase: api.VolumeFailed, Message: "Unknown state"}, nil
+}
+
+// provisionVolume provisions a volume that has been created in the cluster but not yet fulfilled by
+// the storage provider.  This func (and Recycle/Delete) locks on the PV.Name to limit 1 operation per volume at a time.
+func provisionVolume(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+	controller.locks[pv.Name].Lock()
+	defer func(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+		controller.locks[pv.Name].Unlock()
+		delete(controller.locks, pv.Name)
+	}(pv, controller)
+
+	if isAnnotationMatch(pvProvisioningRequired, pvProvisioningCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is already provisioned", pv.Name)
+	}
+
+	qos, exists := pv.Annotations[qosProvisioningKey]
+	if !exists {
+		glog.V(5).Infof("No QoS tier on volume %s.  Provisioning not required.", pv.Name)
+		return
+	}
+
+	provisioner, found := controller.provisionerPlugins[qos]
+	if !found {
+		glog.V(5).Infof("No provisioner found for tier %s", qos)
+		return
+	}
+
+	// Find the claim in local cache
+	obj, exists, _ := controller.claimStore.GetByKey(fmt.Sprintf("%s/%s", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name))
+	if !exists {
+		glog.V(5).Infof("No claim[%s] found for PV[%s]", pv.Spec.ClaimRef.Name, pv.Name)
+		return
+	}
+	claim := obj.(*api.PersistentVolumeClaim)
+
+	creater, _ := newCreater(provisioner, claim)
+	err := creater.Provision(pv)
+	if err != nil {
+		glog.Errorf("Could not provision %s", pv.Name)
+		return
+	}
+
+	oldValue := pv.Annotations[pvProvisioningRequired]
+	pv.Annotations[pvProvisioningRequired] = pvProvisioningCompleted
+	pv, err = controller.client.UpdatePersistentVolume(pv)
+	if err != nil {
+		// rollback
+		pv.Annotations[pvProvisioningRequired] = oldValue
+		// TODO:  https://github.com/kubernetes/kubernetes/issues/14443
+		// the volume was created in the infrastructure and likely has a PV name on it, but we failed to mark the provisioning completed.
+		return
+	}
+}
+
+// recycleVolume recycles a volume that has been released from its claim.
+// This func (and Provision/Delete) locks on the PV.Name to limit 1 operation per volume at a time.
+func recycleVolume(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+	controller.locks[pv.Name].Lock()
```

Comment:
Nit: please log pv.Name on enter, as nearly all log statements in this function do not include it.